src/flint/flint_base/flint_base.pyx

```diff
@@ -114,16 +119,117 @@ cdef class flint_poly(flint_elem):
                 v = - fac[0]
                 roots.append((v, m))
         return roots
-    
+
     def complex_roots(self):
         raise AttributeError("Complex roots are not supported for this polynomial")
 
 
+cdef class flint_mpoly_context(flint_elem):
+    """
+    Base class for multivariate ring contexts
+    """
+
+    _ctx_cache = None
+
+    def __cinit__(self):
+        self._init = False
+
+    def __init__(self, long nvars, names):
+        assert nvars >= 1
+        assert len(names) == nvars
+        self.py_names = tuple(bytes(name, 'utf-8') if not isinstance(name, bytes) else name for name in names)
+        self.c_names = <char**>libc.stdlib.malloc(nvars * sizeof(char *))
+        self._init = True
+        for i in range(nvars):
+            self.c_names[i] = self.py_names[i]
+
+    def __dealloc__(self):
+        if self._init:
+            libc.stdlib.free(self.c_names)
+        self._init = False
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self.nvars()}, '{self.ordering()}', {self.names()})"
+
+    def name(self, long i):
+        assert i >= 0 and i < len(self.py_names)
+        return self.py_names[i].decode('utf-8')
+
+    def names(self):
+        return tuple(name.decode('utf-8') for name in self.py_names)
+
+    def gens(self):
+        return tuple(self.gen(i) for i in range(self.nvars()))
+
+    @staticmethod
+    def create_cache_key(slong nvars, ordering: str, names: str):
+        nametup = tuple(name.strip() for name in names.split(','))
+        if len(nametup) != nvars:
+            if len(nametup) != 1:
+                raise ValueError("Number of variables does not equal number of names")
+            nametup = tuple(nametup[0] + str(i) for i in range(nvars))
+        return nvars, ordering, nametup
+
+    @classmethod
+    def get_context(cls, slong nvars=1, ordering: str = "lex", names: Optional[str] = "x", nametup: Optional[tuple] = None):
```

Comment:
I'm unsure how Python type annotations work with cython...

Do these annotations actually affect the Cython code?

Reply:
![image](https://github.com/flintlib/python-flint/assets/49361082/dd1a91b7-762e-4008-9843-aace336b20d7)
These annotations do affect the cython code, they can enforce the type of arguments, however, since I haven't been consistent with this, I'd be fine with keeping then and typing the rest of the things in this PR, or removing them to be consistent.